<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>shinyGEO by jasdumas</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">shinyGEO</h1>
      <h2 class="project-tagline">a web application for analyzing Gene Expression Omnibus (GEO) datasets using shiny</h2>
      <a href="https://github.com/jasdumas/shinyGEO" class="btn">View on GitHub</a>
      <a href="https://github.com/jasdumas/shinyGEO/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/jasdumas/shinyGEO/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="shinygeo" class="anchor" href="#shinygeo" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>shinyGEO</h1>

<p>shinyGEO is a web-based tool that allows a user to download the expression and sample data from a <a href="http://www.ncbi.nlm.nih.gov/geo/browse/">Gene Expression Omnibus</a> dataset, select a gene of interest, and perform a survival or differential expression analysis using the available data. For both analyses, shinyGEO produces publication-ready graphics and generates R code ensuring that all analyses are reproducible. The tool is developed using shiny, a web application framework for R, a language for statistical computing and graphics.</p>

<h2>
<a id="webpages" class="anchor" href="#webpages" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Webpages</h2>

<p>Primary Application Hosting at Eastern Connecticut State University: <a href="http://bioinformatics.easternct.edu/shinyGEO/">http://bioinformatics.easternct.edu/shinyGEO/</a></p>

<p>Primary Homepage: <a href="http://gdancik.github.io/shinyGEO/">http://gdancik.github.io/shinyGEO/</a></p>

<hr>

<p>Shinyapps.io (Mirror site): <a href="https://jasminedumas.shinyapps.io/shinyGEO">https://jasminedumas.shinyapps.io/shinyGEO</a></p>

<p>Shinyapps.io (Mirror site): <a href="https://gdancik.shinyapps.io/shinyGEO/">https://gdancik.shinyapps.io/shinyGEO/</a></p>

<p>GitHub Repo Page: <a href="http://github.com/jasdumas/shinyGEO/">http://github.com/jasdumas/shinyGEO/</a></p>

<p>Github Homepage: <a href="http://jasdumas.github.io/shinyGEO/">http://jasdumas.github.io/shinyGEO/</a></p>

<h2>
<a id="running-shinygeo-locally" class="anchor" href="#running-shinygeo-locally" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Running shinyGEO locally</h2>

<ol>
<li><p>Download and install <em>R</em> from <a href="https://cran.r-project.org">https://cran.r-project.org</a>. (Note: <em>shinyGEO</em> requires <em>R</em>&gt;=3.23)</p></li>
<li>
<p>Open <em>R</em>. Then type the following (you may copy and paste) into the <em>R</em> console in order to install the CRAN packages required by <em>shinyGEO</em>:</p>

<p><code>install.packages(c("shiny","survival","shinyBS","GGally","ggplot2","shinyAce","knitr","rmarkdown","RCurl","shinyjs","shinydashboard"))</code> </p>

<pre><code>`install.packages('devtools')`
`devtools::install_github('rstudio/DT', ref = "f3e86a6")`
</code></pre>
</li>
<li>
<p>Type the following to install the Bioconductor Packages required by <em>shinyGEO</em>:</p>

<p><code>source("http://bioconductor.org/biocLite.R")</code><br>
<code>biocLite(c("Biobase","GEOquery"))</code></p>
</li>
<li>
<p>Type the following to run <em>shinyGEO</em>:</p>

<p><code>library(shiny)</code><br>
<code>runUrl("https://github.com/gdancik/shinyGEO/archive/master.zip")</code></p>
</li>
</ol>

<h2>
<a id="contributors" class="anchor" href="#contributors" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Contributors</h2>

<p>Main contributors: Jasmine Dumas, Michael Gargano, Garrett M. Dancik, PhD</p>

<p>Additional contributors: Ken-Heng Liao, Gregory Harper, Brandon Spratt</p>

<h2>
<a id="acknowledgements" class="anchor" href="#acknowledgements" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Acknowledgements</h2>

<p>This work was supported, in part, by <a href="http://www.google-melange.com/gsoc/homepage/google/gsoc2015">Google Summer of Code</a> funding to JD in 2015. MG, KHL, GH, and BS contributed as part of an independent study in Computer Science / Bioinformatics while undergraduate students at Eastern Connecticut State University, Willimantic, CT,  USA.</p>

<h2>
<a id="find-out-more" class="anchor" href="#find-out-more" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Find out more</h2>

<p>Feel free to check out the project <a href="https://github.com/jasdumas/shinyGEO/wiki">wiki</a> creating during Google Summer of Code.</p>

<h2>
<a id="updates" class="anchor" href="#updates" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Updates</h2>

<p>25 AUGUST 2015: The current (post-GSoC version) development version as of 21 August 2015 19:00 UTC to the present can be found <a href="https://github.com/jasdumas/shinyGEO">here</a> as per the program code submission guidelines and the submitted GSoC release can be found <a href="https://github.com/jasdumas/shinyGEO/releases/tag/v0.1">here</a>.</p>

<p>11 SEPTEMBER 2015: This GitHub Repo and corresponding links have be updated to a new name to conform with CRAN naming conventions and for deployment guidelines</p>

<p>2 OCTOBER 2015: The application has been deployed to shinyapps.io cloud service at <a href="https://jasminedumas.shinyapps.io/shinyGEO">https://jasminedumas.shinyapps.io/shinyGEO</a> Report bugs, glitches, and general comments as this is the first release designated for testing and configuration.</p>

<p>7 JANUARY 2016: The application will also be deployed on the Eastern CT State University Shiny Server</p>

<p>15 APRIL 2016: The abstract for shinyGEO was accepted to two conferences: <a href="http://www.abstractsonline.com/Plan/ViewAbstract.aspx?mID=4017&amp;sKey=b710c4a6-fafb-4546-a4ef-94ef72d93639&amp;cKey=0243e952-bd00-4008-84b0-53222a594ee9&amp;mKey=1d10d749-4b6a-4ab3-bcd4-f80fb1922267">AACR Annual Meeting 2016 April 16 - 20, 2016 Ernest N. Morial Convention Center New Orleans, Louisiana, USA</a>, and <a href="http://user2016.org/">The R User (useR 2016) Conference June 27 - June 30 2016 Stanford University, Stanford, California</a></p>

<p>...</p>

<h2>
<a id="contact" class="anchor" href="#contact" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Contact</h2>

<p>Jasmine Dumas <a href="https://twitter.com/jasdumas">@jasdumas</a> / <a href="https://twitter.com/shiny_geo">@shiny_GEO</a></p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/jasdumas/shinyGEO">shinyGEO</a> is maintained by <a href="https://github.com/jasdumas">jasdumas</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
